File: src/solutions/python/ImplementTriePrefixTree.py
from typing import Optional, List, Dict
import collections

'''
# Problem
 	- `Link`: https://leetcode.com/problems/implement-trie-prefix-tree/
# Solution
	- `Author`: Kyungtaek Lim (Jonas)
	- `Date`: July 6, 2025
	- `Answer`: Trie / TriePrefixTree
'''


'''
# Option #1
- dict, set
- O(K^2), O(1), O(1)
'''
class Trie:
    def __init__(self):
        self.my_set = set()
        self.my_dict = {}

    # O(k) (k == the length of the word)
    def insert(self, word: str) -> None:
        self.my_set.add(word)
        prefix = ""
        for c in word:
            prefix += c
            self.my_dict[prefix] = True

    # O(1)
    def search(self, word: str) -> bool:
        return word in self.my_set

    # O(1)
    def startsWith(self, prefix: str) -> bool:
        return self.my_dict.get(prefix, False)




'''
# Option #2
- Trie (Prefix Tree)
- O(L), O(L), O(P) (L = The length of the words, P = the length of the prefixes)
- ref) https://www.youtube.com/watch?v=oobqoCJlHA0
'''
class TreeNode:
    def __init__(self):
        self.children = {}
        self.end = False # You cannot tell the end of the tree with only children cause nodes are shared for other words

class TriePrefixTree:

    def __init__(self):
        self.root = TreeNode()

    # O(L) (L = The length of the words)
    def insert(self, word: str) -> None:
        curr = self.root
        for c in word:
            if c not in curr.children:
                curr.children[c] = TreeNode()
            curr = curr.children[c]
        curr.end = True

    # O(L) (L = The length of the words)
    def search(self, word: str) -> bool:
        curr = self.root
        for c in word:
            if c not in curr.children:
                return False
            curr = curr.children[c]
        return curr.end

    # O(P) (P = The length of the prefixes)
    def startsWith(self, prefix: str) -> bool:
        curr = self.root
        for c in prefix:
            if c not in curr.children:
                return False
            curr = curr.children[c]
        return True




'''
# Option #3
- Trie (Prefix Tree)
- O(L), O(W·L), O(W·L) (L = length of the word/prefix you're searching for, W = number of stored words that start with the same first letter as your query (worst case))
- Slower because the root has all the words even though they start with the same alphabet (e.g., 'app' and 'apple' have separate nodes not even partially.)
- September 4, 2026
'''

class Node:
    def __init__(self, val: str, next: Node):
        self.val = val
        self.next = next

class Trie:

    def __init__(self):
        self.map: Dict[str, List[Node]] = {}

    def insert(self, word: str) -> None:
        n: int = len(word)
        node: Optional[Node] = None
        for i in range(n - 1, -1, -1):
            node = Node(word[i], node)
            
        if word[0] not in self.map:
            self.map[word[0]] = []
        self.map[word[0]].append(node)
            

    def search(self, word: str) -> bool:
        if word[0] not in self.map:
            return False
        n: int = len(word)
        for node in self.map[word[0]]:
            node = node.next
            success: bool = True
            for i in range(1, n):
                if not node or node.val != word[i]:
                    success = False
                    break
                node = node.next
            if success and node == None:
                return True
        return False
        

    def startsWith(self, prefix: str) -> bool:
        if prefix[0] not in self.map:
            return False
        n: int = len(prefix)
        for node in self.map[prefix[0]]:
            node = node.next
            success: bool = True
            for i in range(1, n):
                if not node or node.val != prefix[i]:
                    success = False
                    break
                node = node.next
            if success:
                return True
        return False




'''
# Option #4
- Trie (Prefix Tree) - The same as Option #2
- O(L), O(L), O(P) (L = The length of the words, P = the length of the prefixes)
- September 4, 2026
'''
class Node:
    def __init__(self):
        self.next: Dict[str, Node] = {}
        self.end: bool = False

class Trie:

    def __init__(self):
        self.root: Dict[str, Node] = {}

    def insert(self, word: str) -> None:
        n: int = len(word)
        if word[0] not in self.root:
            self.root[word[0]] = Node()
        node: Node = self.root[word[0]]

        for c in word[1:]:
            if c not in node.next.keys():
                node.next[c] = Node()
            node = node.next[c]
            
        node.end = True

    def search(self, word: str) -> bool:
        if word[0] not in self.root:
            return False
            
        node = self.root[word[0]]
        for c in word[1:]:
            if c not in node.next.keys():
                return False
            node = node.next[c]
        return node.end
        

    def startsWith(self, prefix: str) -> bool:
        if prefix[0] not in self.root:
            return False
            
        node = self.root[prefix[0]]
        for c in prefix[1:]:
            if c not in node.next.keys():
                return False
            node = node.next[c]
        return True

    
if __name__ == '__main__':
    word = "1234"
    prefix = "12"
    obj = Trie()
    obj.insert(word)
    param_2 = obj.search(word)
    param_3 = obj.startsWith(prefix)
    
    word = "1234"
    prefix = "12"
    obj = TriePrefixTree()
    obj.insert(word)
    param_2 = obj.search(word)
    param_3 = obj.startsWith(prefix)
    print(param_2, param_3)

File: src/solutions/typescript/ImplementTriePrefixTree.ts

/*
# Problem
 	- `Link`: https://leetcode.com/problems/implement-trie-prefix-tree/
# Solution
	- `Author`: Kyungtaek Lim (Jonas)
	- `Date`: July 18, 2025
	- `Answer`: Trie
*/


/*
# Option #1
- Trie (Prefix Tree)
- O(L), O(L), O(P) (L = The length of the words, P = the length of the prefixes)
- ref) https://www.youtube.com/watch?v=oobqoCJlHA0
*/
class TreeNode { // Might have to change the name of the class (TreeNode -> TreeNodeCustom), due to its duplication issue on LeetCode.

    children: Map<string, TreeNode>;
    end: boolean;

    constructor() {
        this.children = new Map();
        this.end  = false;
    }
}

class Trie {

    node: TreeNode;

    constructor() {
        this.node = new TreeNode();
    }

    insert(word: string): void {
        let curr: TreeNode = this.node;
        for (const c of word) {
            if (!curr.children.has(c)) {
                curr.children.set(c, new TreeNode());
            }
            curr = curr.children.get(c)!;
        }
        curr.end = true;
    }

    search(word: string): boolean {
        let curr: TreeNode = this.node;
        for (const c of word) {
            if (!curr.children.has(c)) return false;
            curr = curr.children.get(c)!;
        }
        return curr.end;
    }

    startsWith(prefix: string): boolean {
        let curr: TreeNode = this.node;
        for (const c of prefix) {
            if (!curr.children.has(c)) return false;
            curr = curr.children.get(c)!;
        }
        return true;
    }
}




/*
# Option #2
- Trie (Prefix Tree)
- O(L), O(W·L), O(W·L) (L = length of the word/prefix you're searching for, W = number of stored words that start with the same first letter as your query (worst case))
- Slower because the root has all the words even though they start with the same alphabet (e.g., 'app' and 'apple' have separate nodes not even partially.)
- September 4, 2026
*/
class _Node {
    val: string
    next: _Node | null

    constructor(val: string, next?: _Node | null) {
        this.val = val;
        this.next = next === undefined ? null : next;
    }
}


class Trie {
    dict: Map<string, Array<_Node>>

    constructor() {
        this.dict = new Map();
    }

    insert(word: string): void {
        const n: number = word.length;
        let node: _Node | null = null;
        for (let i = n - 1; i >= 0; i--) {
            node = new _Node(word[i], node);
        }
        if (!this.dict.has(word[0])) this.dict.set(word[0], []);
        this.dict.get(word[0])!.push(node!);
    }

    search(word: string): boolean {
        const n: number = word.length;
        if (!this.dict.has(word[0])) return false;
        
        for (const curr of this.dict.get(word[0])!) {
            let node: _Node | null = curr.next;
            let success: boolean = true;
            for (let i = 1; i < n; i++) {
                if (node === null || node.val !== word[i]) {
                    success = false;
                    break;
                }
                node = node.next;
            }
            if (success && node === null) return true;
        }
        return false;
    }

    startsWith(prefix: string): boolean {
        const n: number = prefix.length;
        if (!this.dict.has(prefix[0])) return false;
        
        for (const curr of this.dict.get(prefix[0])!) {
            let node: _Node | null = curr.next;
            let success: boolean = true;
            for (let i = 1; i < n; i++) {
                if (node === null || node.val !== prefix[i]) {
                    success = false;
                    break;
                }
                node = node.next;
            }
            if (success) return true;
        }
        return false;
    }
}




/*
# Option #3
- Trie (Prefix Tree) - The same as Option #1
- O(L), O(L), O(P) (L = The length of the words, P = the length of the prefixes)
- September 4, 2026
*/
class _Node {
    next: Map<string, _Node>
    end: boolean

    constructor() {
        this.next = new Map()
        this.end = false;
    }
}


class Trie {
    root: Map<string, _Node>

    constructor() {
        this.root = new Map();
    }

    insert(word: string): void {
        const n: number = word.length;
        if (!this.root.has(word[0])) this.root.set(word[0], new _Node());
        let node: _Node = this.root.get(word[0])!;
        
        for (let i = 1; i < n; i++) {
            if (!node.next.has(word[i])) node.next.set(word[i], new _Node());
            node = node.next.get(word[i]);
        }

        node.end = true;
    }

    search(word: string): boolean {
        const n: number = word.length;
        if (!this.root.has(word[0])) return false;

        let node: _Node = this.root.get(word[0]);
        for (let i = 1; i < n; i++) {
            if (!node.next.has(word[i])) return false;
            node = node.next.get(word[i]);
        }
        return node.end;
    }

    startsWith(prefix: string): boolean {
        const n: number = prefix.length;
        if (!this.root.has(prefix[0])) return false;

        let node: _Node = this.root.get(prefix[0]);
        for (let i = 1; i < n; i++) {
            if (!node.next.has(prefix[i])) return false;
            node = node.next.get(prefix[i]);
        }
        return true;
    }
}